library(shiny)
library(shinydashboard)
library("readxl")
library(ggplot2)
library(DT)
library(Rsamtools)

options(shiny.maxRequestSize=500*1024^2)

# SvCov version 2.0 (06-07-2021)
# Create Bin, scan coverage and filter repeated sequences and use bamfile to find breaks in coverage
# Use with R studio
# Input are fichier tsv generate with bedtools as follows
# Needs also a bamfile generated with samtools using the same reference genome used for bedtools
# first align on ref genome using minimap2 : minimap2 -ax map-ont refgenome.fa fichier.fq | samtools sort -o fichier.bam)
# index ref genome : samtools faidx
# generate bedfile : bedtools genomecov -d -ibam ./fichier.bam > fichier.tsv
# "la patience est un plat qui se mange sans sauce" Perceval


########################
# Define the functions #
########################


# filter repeated sequences at base pair level

filter <- function(x,y)
{
  # prepare data
  x[,4]=paste(x[,1], x[,2], sep = "_")
  y[,3]=paste(y[,1], y[,2], sep = "_")

  #compute differencies and filter
  valid=setdiff(x[,4],y[,3])
  xf=x[x[,4] %in% valid,]  
  return(xf)
}

# filter output file based on bam coordinates (utiliser la fonction select de dplyr)

FilterCut <- function(x,y,w)
{
  l=dim(x)[1]
  
  res=data.frame("ID","chr",1)
  colnames(res)=c("ID","chr","pos")
  for (i in 1:l)
  {
    temp=x[i,]
    nb=2*w+1
    start=x[i,2]-w
    end=x[i,2]+w
    c=rep(paste(x[i,1],x[i,2],sep="_"),nb)
    a=rep(x[i,1],nb)
    b=rep(start:end)
    tab=data.frame(c,a,b)
    colnames(tab)=c("ID","chr","pos")
    top=rbind(res,tab)
    res=top
  }
  
  for (i in 1:l)
  {
    temp=x[i,]
    nb=2*w+1
    start2=x[i,3]-w
    end2=x[i,3]+w
    c=rep(paste(x[i,1],x[i,2],sep="_"),nb)
    a=rep(x[i,1],nb)
    b=rep(start2:end2)
    tab=data.frame(c,a,b)
    colnames(tab)=c("ID","chr","pos")
    top=rbind(res,tab)
    res=top
  }
  
  res[,4]=paste(res[,2],res[,3],sep="_") 
  y[,4]=paste(y[,1],y[,2],sep="_")
  
  valid=intersect(res[,4],y[,4])
  resf=res[res[,4] %in% valid,]
  
  x[,6]=paste(x[,1],x[,2],sep = "_")
  valid2=unique(resf[,1])
  xf=x[x[,6] %in% valid2 ,]
  xf=xf[,1:5]
  
  return(xf)
}


# create bin from tsv file
##########################

Bining <- function(x,y,z,w)
{
  # calculer les bin dans le tableau d'origine
  bin=round(y/w,0)*w
  win=paste(x, bin, sep = "_")
  
  # Faire le calcul des mÃ©dianes par bin (aggregation)
  m=tapply(z, win, median)

  # Formatage et tri du tableau
  m=data.frame(key=names(m), value=m)
  m[,1]=as.character(m[,1])
  for (i in 1:length(m[,1]))
  {
    m[i,3]=unlist(strsplit(m[i,1],"_"))[1]
    m[i,4]=unlist(strsplit(m[i,1],"_"))[2]
  }
  m[,5]=m[,2]
  m=m[,3:5]
  m[,2]=as.numeric(m[,2])
  m=m[order(m[,1],m[,2]),]
  
  # Valeurs pour tracer les chromosomes
  index=c(1,rep(NA,length(m[,1])-1))
  m[,4]=index
  for (j in 2:length(m[,1]))
  {
    if(m[j,1]!=m[j-1,1] && m[j-1,4]==1) {m[j,4]=3}
    else if(m[j,1]!=m[j-1,1] && m[j-1,4]==3) {m[j,4]=1}
    else {m[j,4]=m[j-1,4]}
  }
  return(m)
}


# Find blocs of DEL and DUP
###########################

CNV <- function(x,u,d,w)
# x is the table of coverage generated by the previous function, x and d are the threshold defined with the interface and w the window size  
{
  # prepare data
  med=median(x[,3])
  x[,5]=rep("NA",length(x[,1]))
  lower=d*med
  upper=u*med
  
  # find window up and down
  for (i in 1:length(x[,1]))
    {
     if(x[i,3]<lower){x[i,5]="DEL"}
     else if(x[i,3]>upper){x[i,5]="DUP"}
     else {x[i,5]="none"}
    }

  # find window up and down
  CNV=subset(x,x[,5]!="none")    
  index=c(1,rep(NA,length(CNV[,1])-1))
  CNV[,6]=index
  for (j in 2:length(CNV[,1]))
  {
    if(CNV[j,1]==CNV[j-1,1] && CNV[j,2]-CNV[j-1,2]>2*w) {CNV[j,6]=CNV[j-1,6]+1}
    else if (CNV[j,1]==CNV[j-1,1] && CNV[j,2]-CNV[j-1,2]<=2*w) {CNV[j,6]=CNV[j-1,6]}
    else {CNV[j,6]=1}  
  }
  CNV[,7]=paste(CNV[,1],CNV[,6], sep = "_")

  # extract bloc data
  lev=unique(CNV[,7])
  res=matrix(nrow=length(lev), ncol=4)
  res=as.data.frame(res)
  for(l in 1:length(lev))
  {
    temp=subset(CNV,CNV[,7]==lev[l])
    min=min(temp[,2])
    max=max(temp[,2])    
    res[l,1]=as.character(lev[l])
    res[l,2]=min
    res[l,3]=max
    res[l,4]=as.character(temp[1,5])
  }
  res[,5]=(res[,3]-res[,2])+w
  for (q in 1:length(res[,1]))
  {
    res[q,1]=unlist(strsplit(res[q,1],"_"))[1]
  }
  res=data.frame(ID=res[,1], start=res[,2], end=res[,3], length=res[,5], Type=res[,4])
  return(res)
}


# Prepare data for plotting
###########################

xaxt <- function(x) 
{
  x[,5]=rep(1,length(x[,1]))
  for (i in 2:length(x[,1]))
  {
    if(x[i,1]==x[i-1,1]) {x[i,5]=x[i-1,5]+(x[i,2]-x[i-1,2])}
    else if(x[i,1]!=x[i-1,1]){x[i,5]=x[i-1,5]+(x[i,2])}
  }
  return(x)
}


# Calculate stat par chromosomes
################################

STAT <- function(x)
{
  lev=unique(x[,1])
  res=matrix(nrow=length(lev), ncol=2)
  res=as.data.frame(res)
  for(i in 1:length(lev))
  {
    temp=subset(x,x[,1]==lev[i])
    med=median(as.numeric(temp[,3]))
    res[i,1]=as.character(lev[i])
    res[i,2]=med  
  }
  res=data.frame(Chr=res[,1], MedianCov=res[,2])
  return(res)
}


# Import and Process BamFile
############################

Bam <- function(x)
{
  a=x[[1]]$rname
  b=x[[1]]$pos  
  z=cbind(a,b)
  z=as.data.frame(z)
  z$c=paste(a,b,sep="_")
  z=subset(z,z$a!="NA")  
  s=table(z$c)
  s=as.data.frame(s)
  r=subset(s,s$Freq>3)
  r[,1]=as.character(r[,1])
  for (i in 1:length(r[,1]))
  {
    r[i,3]=unlist(strsplit(r[i,1],"_"))[1]
    r[i,4]=unlist(strsplit(r[i,1],"_"))[2]
  }
  r[,5]=r[,2]
  r=r[,3:5]
  return(r)
}


#################################
#    Define User interface      #
#################################

ui <- dashboardPage(skin = "red",
                    dashboardHeader(title = " CASPIAN 2.0 : CoverAge Structural variation graPhIcal ANalysis ", titleWidth = 700), 
                    dashboardSidebar(width=250,
                                     box(width=250, background="green",
                                         fileInput("file1","Choose TSV file to upload"),
                                         fileInput("file2","Choose filters file"),
                                         fileInput("file3","Choose BAM file")
                                     ),
                                     box(width = 250, background="blue",
                                         numericInput("Wd","Choose window size (pb)", 1000, min = 1000, max = 100000),
                                         sliderInput("slider","Choose Lower and Upper limits",min = 0, max = 5, value = c(0.2,1.5), step = 0.1, ticks = FALSE),
                                         tags$hr(),
                                         actionButton("cal", "Update Plot and Analysis"),
                                         tags$hr(),
                                         selectInput("Chr","Select a chromosome", choices=""),
                                     ),
                                     box(width = 250, background="red",
                                         downloadButton("saveR","  Save results in a txt file  ", class = "butt"), tags$head(tags$style(".butt{background-color:black;} .butt{color:black;}")),
                                         tags$hr(),
                                         downloadButton("saveG","  Export genome Plot  ", class = "butt")
                                         )
                    ),
                    dashboardBody(
                         fluidRow(
                           column(width = 12,
                                  box(width = 10,
                                    plotOutput("genome"), status="success"
                                  ),
                                  box(width = 2,
                                    verbatimTextOutput("stats")
                                  ),
                                 ),
                         ),  
                        fluidRow(
                          column(width = 12,
                                 box(width = 8,
                                   plotOutput("chromo"), status="success"
                                 ),
                                 box(width = 4,
                                   dataTableOutput("Gtable"), status ="danger"
                                 )
                                 )                          
                        )                     
                    )
)


#################################
#    Define Server action       # 
#################################

server <- function(session, input, output){

# Import et format data  
  
  data <- reactive({
    inFile <- input$file1
    if (is.null(inFile))
       return(NULL)
    as.data.frame(read.table(inFile$datapath))     
  })
  
  filtre <- reactive({
    inFile2 <- input$file2
    if(is.null(inFile2))
      return(NULL)
    as.data.frame(read.table(inFile2$datapath, header = T))
  })
  
  BAM <- reactive({
    inFile3 <- input$file3
    if(is.null(inFile3))
      return(NULL)
    scanBam(inFile3$datapath)     
  })
  
  dataf <-  reactive({
    filter(data(),filtre())
  })

  Chromosome <- reactive({
    unique(dataf()[,1])
  })
  
  observe({
    updateSelectInput(session, inputId = "Chr", choices = Chromosome())  
  })
  

# Transform the data : calculate median on each bin and find blocs
  
  bin <- eventReactive(input$cal, {
    Bining(dataf()[,1],dataf()[,2],dataf()[,3],input$Wd)
  })

  DupDel <- eventReactive(input$cal, {
     CNV(bin(),input$slider[2],input$slider[1],input$Wd)
  })
  
  DupDelFilt <- eventReactive(input$cal, {
    FilterCut(DupDel(),Cut(),input$Wd)
  })
  
  output$Gtable <-renderDataTable(
     DupDel(),options = list(pageLength = 6)
#     DupDelFilt(), options = list(pageLength = 7)
  )

  Cut <- reactive({
    Bam(BAM())
  })
  
# Plot coverage  
  
  bing<- reactive({
    xaxt(bin()) 
  })
  
  output$genome <- renderPlot({
  inFile <- input$file3
  if (is.null(inFile))
    return(NULL)
  plot(bing()[,5],bing()[,3]/median(bing()[,3]), ylim=c(-1,5), xlab = "coord", ylab = "coverage / median(coverage)", main = "relative genome wide coverage", pch = 20,type = "b")
  par(new = T)
  plot(bing()[,5],bing()[,4]/-10, ylim=c(-1,5), xlab = "", ylab = "", main = "", pch = 20, col = "blue", cex = 0.5)
  abline(h=input$slider[2], col = "red", lty=2)
  abline(h=input$slider[1], col = "red", lty=2)
  
  })  

# Calculate chromosome stats and print it  
  
  ChrStat <- reactive({
    STAT(bin())
  })
  
  output$stats <- renderPrint({
    inFile <- input$file3
    if (is.null(inFile))
      return(NULL)
    ChrStat()
  })
  
  
# Plot particular chromosome  
   
  chrplot <- reactive({
  as.data.frame(subset(bing(),bing()[,1]==input$Chr))  
  }) 
  
  output$chromo <- renderPlot({
    inFile <- input$file3
    if (is.null(inFile))
      return(NULL)
    plot(chrplot()[,2], chrplot()[,3], ylim=c(0,4*median(chrplot()[,3])), xlab = "coord", ylab = "coverage", main = as.character(input$Chr), pch = 20)
    abline(h=input$slider[2]*median(bin()[,3]), col = "red", lty=2)
    abline(h=input$slider[1]*median(bin()[,3]), col = "red", lty=2)
  })  

# Export data and graph
  
  output$saveR <- downloadHandler(
    filename = "Res.txt",
    content = function(file) {
    write.table(DupDelFilt(), file, row.names = F, sep = "\t", quote = FALSE)
    })

  exportPlot <- reactive({
    plot(bing()[,5],bing()[,3]/median(bing()[,3]), ylim=c(-1,5), xlab = "coord", ylab = "coverage / median(coverage)", main = "relative genome wide coverage", pch = 20,type = "b")
    par(new = T)
    plot(bing()[,5],bing()[,4]/-10, ylim=c(-1,5), xlab = "", ylab = "", main = "", pch = 20, col = "blue", cex = 0.5)
    abline(h=input$slider[2], col = "red", lty=2)
    abline(h=input$slider[1], col = "red", lty=2)
  })
  
  output$saveG <- downloadHandler(
  filename ="GenomeCov.pdf",  
  content = function(file) {
    ggsave(file,exportPlot(), width = 20, height = 5, device = "pdf")
  })

}


#################################
#        Launch in local        # 
#################################

shinyApp(ui = ui, server = server)
